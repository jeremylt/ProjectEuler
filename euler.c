#include <stdio.h>
#include "euler.h"
#include "utils.h"

// -----------------------------------------------------------------------------
// Driver
// -----------------------------------------------------------------------------
int main(void) {
  int menuOption = 0;

  printf("-------------------------------------------------------------------\n"
         "  Welcome to a Project Euler solver:\n"
         "-------------------------------------------------------------------\n"
         "  Menu Option 1: Print help information\n"
         "  Menu Option 2: Solve Euler problem\n"
         "  Menu Option 3: Exit\n\n");

  // Loop until exit
  while (menuOption != 3) {
    // Prompt user
    getValidUserInt("  Enter a menu option (1-3): ", stdin, 1, 3, &menuOption);

    // Take selected action
    switch (menuOption) {
    case 1:
      printf("---------------------------------------------------------------\n"
             "  Help:\n"
             "    Menu Option 1: Print help information\n"
             "    Menu Option 2: Solve Euler problem\n"
             "    Menu Option 3: Exit\n\n");
      break;

    case 2: {
      int problemOption;

      printf("---------------------------------------------------------------\n"
             "  Euler Problems 1-%d are avaliable\n", NUM_PROBLEMS);
      getValidUserInt("  Select an Euler problem to solve: ", stdin, 1,
                      NUM_PROBLEMS, &problemOption);

      (*problems[problemOption-1])(stdin);
    } break;

    default :
      printf("---------------------------------------------------------------\n"
             "  Exit\n\n"
             "---------------------------------------------------------------\n"
            );
      break;
    }
  }

  return 0;
}
// -----------------------------------------------------------------------------
