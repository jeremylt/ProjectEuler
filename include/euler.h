#ifndef EULER_H
#define EULER_H
#include <math.h>

#define NUM_PROBLEMS 3

// -----------------------------------------------------------------------------
// Solution code
// -----------------------------------------------------------------------------
int problem1(int a, int b, int limit, int *sum);
int problem2(int a, int limit, int *sum);
int problem3(long number, long *factor);

// -----------------------------------------------------------------------------
// Command line interfaces
// -----------------------------------------------------------------------------
typedef int funcPtr(void);
funcPtr problem1CL, problem2CL, problem3CL;

funcPtr *problems[NUM_PROBLEMS] = {
  problem1CL,
  problem2CL,
  problem3CL
};

#endif
