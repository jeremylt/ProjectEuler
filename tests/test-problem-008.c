/// @file
/// Test Project Euler Problem 8
/// \test Test problem 8 result
#include <stdio.h>
#include "euler.h"
// -----------------------------------------------------------------------------
int main(void) {
  FILE *stream;

  // Open file
  stream = fopen("tests/input/test-problem-008.in", "r");
  if(stream == NULL) {
    // LCOV_EXCL_START
    perror("Error opening file\n");
    return(-1);
    // LCOV_EXCL_STOP
  }

  // Get the biggest 4 digit product
  problem008CL(stream);

  // Get the biggest 13 digit product
  problem008CL(stream);

  // Close
  fclose(stream);

  return 0;
}
// -----------------------------------------------------------------------------
