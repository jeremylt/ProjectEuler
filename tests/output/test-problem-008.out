    This code finds the greatest product of [n] adjacent digits in
      the 1000 digit number in data/1000digit.txt

      Enter n:     The greatest product of 4 consecutive digits is: 5832

    This code finds the greatest product of [n] adjacent digits in
      the 1000 digit number in data/1000digit.txt

      Enter n:     The greatest product of 13 consecutive digits is: 23514624000

